### PR TITLE
[Fixes #53] Implement copy for gpkg layers

### LINKED ISSUE #53: Copy / clone resource should clone also the dynamic model

### PR DESCRIPTION
**Pull request by mattiagiupponi**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -20,19 +20,16 @@ CELERY_TASK_QUEUES += (
...
+    Queue('importer.copy_dynamic_model', GEONODE_EXCHANGE, routing_key='importer.copy_dynamic_model'),
+    Queue('importer.copy_geonode_data_table', GEONODE_EXCHANGE, routing_key='importer.copy_geonode_data_table'),
+)
 )
 
 DATABASE_ROUTERS = ["importer.db_router.DatastoreRouter"]
 
 SIZE_RESTRICTED_FILE_UPLOAD_ELEGIBLE_URL_NAMES += ('importer_upload',)
 
-ADDITIONAL_DATASET_FILE_TYPES = [{
```
Comment (mattiagiupponi) [inline README.md:29]: is part of the app setup, so is not needed to force it into the settings anymore

File: importer/handlers/gpkg/handler.py [hunk 2]
```diff
@@ -373,19 +394,6 @@ def _update_execution_request(self, execution_id: str, **kwargs):
             status=ExecutionRequest.STATUS_RUNNING, **kwargs
         )
 
-    def _create_alternate(self, layer_name, execution_id):
```
Comment (mattiagiupponi) [inline importer/handlers/gpkg/handler.py:376]: extracted as common utility method

File: importer/handlers/gpkg/handler.py [hunk 3]
```diff
@@ -525,3 +533,82 @@ def error_callback(*args, **kwargs):
...
+            with connections[db_name].cursor() as cursor:
+                cursor.execute(
+                    f"CREATE TABLE {new_dataset_alternate} AS TABLE {original_dataset_alternate};"
+                )
+
+        task_params = ({}, exec_id, handlers_module_path, actual_step, layer_name, alternate, action)
+
+        import_orchestrator.apply_async(task_params, kwargs)
+    except Exception as e:
+        raise InvalidGeopackageException(detail=e)
```
Comment (giohappy) [inline importer/handlers/gpkg/handler.py:613]: @mattiagiupponi is `InvalidGeopackageException` the rigth exception here?

File: importer/celery_tasks.py [hunk 4]
```diff
@@ -1,25 +1,27 @@
...
 from celery import Task
 from django.utils import timezone
 from django.utils.module_loading import import_string
-
+from django.utils.translation import ugettext
 from importer.api.exception import (InvalidInputFileException,
-                                    PublishResourceException,
+                                    PublishResourceException, ResourceCopyException,
                                     ResourceCreationException,
                                     StartImportException)
```
Comment (etj) [inline importer/celery_tasks.py:12]: We may want to use consistent naming:
e.g. ACTIONResourceException or ResourceACTIONException (not a mix of the two)

File: importer/handlers/gpkg/handler.py [hunk 5]
```diff
@@ -115,7 +129,17 @@ def extract_params_from_data(_data):
         }, _data
 
     @staticmethod
-    def extract_resource_to_publish(files, layer_name, alternate):
+    def extract_resource_to_publish(files, action, layer_name, alternate):
+        if action == 'copy':
```
Comment (etj) [inline importer/handlers/gpkg/handler.py:133]: Use constants

File: importer/handlers/utils.py [hunk 6]
```diff
@@ -24,3 +26,18 @@ def should_be_imported(layer: str, user: get_user_model(), **kwargs) -> bool:
...
+def create_alternate(layer_name, execution_id):
+    '''
+    Utility to generate the expected alternate for the resource
+    is alternate = layer_name_ + md5(layer_name + uuid)
+    '''
+    _hash = hashlib.md5(
+        f"{layer_name}_{execution_id}".encode('utf-8')
+    ).hexdigest()
+    alternate = f"{layer_name}_{_hash}"
+    if len(alternate) >= 64: # 64 is the max table lengh in postgres
```
Comment (etj) [inline importer/handlers/utils.py:40]: if 64 is ok, let's check only for ">", not ">="

Comment (mattiagiupponi) [reply importer/handlers/utils.py:40]: the max lengh is 63, so or we keep this or it should be changed into: >63

**Review by giohappy — APPROVED** [review]

**Pull request merged**